Add support for multimodal content and VLM evaluation

teetone opened the pull request:
## Changes

- Support multimodal content in HELM (see below)
- Added [Visual Question Answering (VQA)](https://visualqa.org) scenario
- Added[ IDEFICS models](https://huggingface.co/HuggingFaceM4/idefics-80b-instruct): IDEFICS 9B/80B and instruct 9B/80B

## Support for multimodal content (updated 10/8/2023)

- `MediaObject`
  - `mime_type: str` - a valid multi-purpose internet mail extension
  - `location: str = ""` - can be a local path or URL
  - `text: str = ""` - when the media object is text
  - `is_local_file: bool`
    - Autocomputed by checking if `location` is a URL
  - On instantiation, we check that the location/path is valid if it's a local file
  - media_object.py will have helper functions to combine lists of `MediaObject`s
- `MultimediaObject`
  - `content: List[MediaObject]`
- `Input` of `Instance`
  - `text: str = ""`
  - `multimedia_content: Optional[MultimediaObject] = None`
    - A single instance can now have interleaved text and files (e.g., images for VLMs)
- `Output` of `Reference`
  - `text: str = ""`
  - `multimedia_content: Optional[MultimediaObject] = None` 
- `InContextLearningMultimodalAdapter` (abstract class)
  - Takes train and eval multimodal contents and `AdapterSpec` parameters to generate `MultimodalPrompt`
  - Uses `WindowService` to remove train instances, so final `MultimodalPrompt` fits within max context length
  - Examples of child classes:
    - `MultimodalGenerationAdapter`
    - `MultimodalMultipleChoiceSeparateAdapter`
    -  ...
- `MultimodalPrompt`: this is the multimodal equivalent of `Prompt` that our text adapters work with
  - `global_prefix: str`
  - `instance_prefix: str`
  - `instructions: str`
  - `train_instance_blocks: List[MultimediaObject]`
  - `eval_instance_blocks: MultimediaObject`
  - Contains train multimodal content and eval multimodal content
  - Generates final  `MultimediaObject` via `multimodal_object` to be sent to the model via `Request`
- `Request`
  - `model: str`
  - `max_tokens: int`
  - `prompt: str = ""`
  - `multimodal_prompt: Optional[MultimediaObject] = None`
    - Multimodal content interleaved (e.g., text, image, text, image, etc. )
  - ...

## Example

Example in-context learning prompt and output of IDEFICS instruct 9B on VQA:

User:
<img src="https://github.com/stanford-crfm/helm/assets/16793796/e3471297-e341-4bf6-a79f-97193b332929" width="200" />
Is this an office desk?<end_of_utterance>

Assistant: yes<end_of_utterance>

User:
<img src="https://github.com/stanford-crfm/helm/assets/16793796/7b65e6a0-cb02-4688-850c-36d531a45240" width="200" />
Is this cat sleeping?<end_of_utterance>

Assistant: no<end_of_utterance>

User: 
<img src="https://github.com/stanford-crfm/helm/assets/16793796/75c258c5-6c58-494b-8c9b-4b9bd01c6daf" width="200" />
How many monitors are on the desk?<end_of_utterance>

Assistant: 1<end_of_utterance>

User: 
<img src="https://github.com/stanford-crfm/helm/assets/16793796/9ef08614-47e7-450e-bfeb-3c7681ec25ac" width="200" />
Where is the horse?<end_of_utterance>

Assistant: _**The horse is on the street.**_

yifanmai commented on src/helm/common/request.py:
```diff
@@ -59,6 +60,9 @@ class Request:
     if messages is specified for a chat model, the prompt is ignored.
     Otherwise, the client should convert the prompt into a message."""
 
+    multimodal_prompt: Optional[MultimodalContent] = None
```
This seems very confusing because I would expect the type to be `multimodal_prompt: Optional[MultimodalPrompt]` but  `MultimodalPrompt` is not `MultimodalContent`...

I feel like `MultimodalPrompt` should be named something else.

teetone replied:
We discussed this offline. Decided to keep for now.

yifanmai commented on src/helm/benchmark/window_services/window_service_factory.py:
```diff
@@ -285,6 +291,14 @@ def get_window_service(model_name: str, service: TokenizerService) -> WindowServ
 
             window_service = LitGPTWindowServce(service)
+        elif model_name == "HuggingFaceM4/idefics-9b":
+            window_service = IDEFICS9bWindowService(service)
```
You can just do `window_service = HuggingFaceWindowService(None, HuggingFaceHubModelConfig.from_string("HuggingFaceM4/idefics-9b"))`

yifanmai commented on src/helm/benchmark/window_services/idefics_window_service.py:
```diff
@@ -0,0 +1,49 @@
+from .tokenizer_service import TokenizerService
+
+
+class IDEFICSWindowService(LocalWindowService, ABC):
```
Don't need this file; just use `HuggingFaceWindowService`

yifanmai reviewed (changes requested):
Thanks! Leaving high level questions; we can discuss more offline.

yifanmai commented:
cc @JosselinSomervilleRoberts

teetone commented on src/helm/proxy/clients/vision_language/idefics_client.py:
```diff
@@ -0,0 +1,163 @@
+from helm.proxy.clients.client import Client, wrap_request_time, cleanup_tokens
+
+
+class IDEFICSClient(Client):
```
I will follow up on this once we add more models.

teetone commented on src/helm/benchmark/run_specs.py:
```diff
@@ -781,6 +809,40 @@ def wrap(func: F) -> F:
+    # TODO: finalize the adapter spec parameters
+    adapter_spec: AdapterSpec = get_vlm_generation_adapter_spec(
+        input_prefix="User: ",
+        input_suffix="<end_of_utterance>",
+        output_prefix="\nAssistant: ",
+        output_suffix="<end_of_utterance>",
+        max_train_instances=3,
+        stop_sequences=["<end_of_utterance>"],
```
This is model-specific but I will finalize the `AdapterSpec` parameters once we add more models. I added a TODO for now.

percyliang commented on src/helm/benchmark/adaptation/adapters/in_context_learning_adapter.py:
```diff
@@ -224,16 +224,28 @@ def construct_example_prompt(self, instance: Instance, include_output: bool, ref
 
-        # References (optionally) and output
-        output: str
+        if include_output:
```
It'd be helpful if you could go through (for next time) and add some general guiding comments like "this piece of code is moved up from below"

teetone replied:
Ah, the diff was a bit confusing. Yes, I will make sure to do that next time.

percyliang commented on src/helm/benchmark/adaptation/adapters/multimodal/in_context_learning_multimodal_adapter.py:
```diff
@@ -0,0 +1,141 @@
+
+    def _fit_multimodal_prompt(self, prompt: MultimodalPrompt) -> MultimodalPrompt:
+        """
+        The prompt consists of instructions, training instances, and the evaluation input.
```
Do we need need all this logic for the multimodal case?  I don't expect the text to be super long.  I wonder if we should be starting to shed some of the complexity from original HELM...  Are we thinking about primarily doing in-context learning or zero-shot for VLMs?

teetone replied:
More models are starting to support in-context learning, so I think we should support both in-context learning and zero-shot. For scenarios like VQA, the text is short, but there could be scenarios in the future where the text is long,  which this adapter is equipped to handle.

percyliang commented on src/helm/benchmark/adaptation/adapters/multimodal/multimodal_prompt.py:
```diff
@@ -0,0 +1,52 @@
+    instructions_block: str
+
+    # Train instance blocks for the prompt
+    train_instance_blocks: List[List[MediaObject]]
```
I think we should abstract `List[MediaObject]` into a separate dataclass because it shows up multiple times and it is semantically a thing - maybe `MultimediaObject` or something?

teetone replied:
That makes sense. I had `MultimediaContent` before, but I like the name `MultimediaObject` to be consistent with `MediaObject`.

teetone replied:
I updated the code and PR description above.

percyliang commented on src/helm/benchmark/run_specs.py:
```diff
@@ -403,6 +404,33 @@ def get_machine_translation_adapter_spec(
     )
 
 
+def get_vlm_generation_adapter_spec(
```
We should really break up this file into multiple ones - can we start doing that with this PR?

percyliang commented on src/helm/benchmark/scenarios/vision_language/vqa_scenario.py:
```diff
@@ -0,0 +1,122 @@
+    Visual Question Answering (VQA) is a dataset containing open-ended questions about images.
+    These questions require an understanding of vision, language and commonsense knowledge to answer.
+
+    We use the v2.0 version released in 2017:
```
I'd include the title and author names.

percyliang commented on src/helm/proxy/clients/vision_language/idefics_client.py:
```diff
@@ -0,0 +1,178 @@
+from helm.proxy.clients.client import Client, wrap_request_time, cleanup_tokens, generate_uid_for_multimodal_prompt
+
+_models_lock: Lock = Lock()
+_models: Dict[str, Tuple[Optional[IdeficsForVisionText2Text], Optional[AutoProcessor]]] = {
```
Should define a dataclass...tuples are semantically impoverished...

percyliang commented on src/helm/benchmark/scenarios/scenario.py:
```diff
@@ -55,7 +56,11 @@ class Input:
     """
 
-    text: str
+    text: str = ""
```
Semantically, it would be more accurate if `text` were `Optional[str] = None`

teetone replied:
I tried to update it to `Optional[str]`, but type checking breaks with `40 errors in 20 files`.

percyliang commented on src/helm/benchmark/scenarios/scenario.py:
```diff
@@ -55,7 +56,11 @@ class Input:
+    text: str = ""
+    """The text of the input (e.g, passage to summarize, text for sentiment analysis, etc.)"""
+
+    multimodal_content: Optional[MultimediaObject] = None
```
Should make terms consistent - object versus content, multimedia versus multimodal...  how about just `multimedia_object` or at least `multimedia_content`?

teetone replied:
I changed it to `multimedia_content` and updated the PR description with the new names. I left `multimodal` for objects on the Adapter level: `MultimodalPrompt`, `multimodal_prompt` in `Request`, `InContextLearningMultimodalAdapter`, etc. (which seems to make more sense), or do you think it should all be `multimedia`?

percyliang commented on src/helm/benchmark/scenarios/scenario.py:
```diff
@@ -82,6 +87,10 @@ class Output:
     """
 
     text: str
+    """The text of the output."""
```
Same comments as for `Input` - do we actually need two classes for this or could they be one?

teetone replied:
Do you mean for `Input` and `Output`?

yifanmai commented on src/helm/proxy/clients/vision_language/idefics_client.py:
```diff
@@ -0,0 +1,194 @@
+            "max_length": request.max_tokens,
+            "bad_words_ids": processor.tokenizer(self.BAD_WORD_TOKENS, add_special_tokens=False).input_ids,
+        }
+        if self.END_OF_UTTERANCE_TOKEN in request.stop_sequences:
```
Please add a comment that explains what this does.

yifanmai commented on src/helm/proxy/clients/client.py:
```diff
@@ -146,3 +147,13 @@ def cleanup_tokens(tokens: List[str], tokenizer_name: Optional[str] = None) -> L
+    """Generates a unique identifier for a given multimodal prompt."""
+    return "".join(
+        [
+            media_object.text if media_object.is_type(TEXT_TYPE) and media_object.text else str(media_object.location)
```
No action needed: This assumes that the file is static and does not get modified, which should be true in most cases... otherwise we'd need a last modified time and / or checksum.

teetone replied:
Yes, I think it's true for most cases. If a user needs to modify `MediaObject`, then I think it would be a `file/MediaObject`.

yifanmai commented on src/helm/common/images_utils.py:
```diff
@@ -0,0 +1,42 @@
+    """Returns the base64 representation of an image file."""
+    image_file = io.BytesIO()
+    image: Image = open_image(image_location)
+    image.save(image_file, format="JPEG")
```
1. Consider taking format as a parameter
2. Consider using PNG, which is usually lossless (JPEG is usually lossy).

yifanmai reviewed (changes requested):
Partial review (sorry, didn't have time to get through all of this today).

teetone commented on setup.cfg:
```diff
@@ -56,7 +56,8 @@ install_requires=
 
     # Models and Metrics Extras
-    transformers~=4.28.1  # For anthropic_client, huggingface_client, huggingface_tokenizer, test_openai_token_cost_estimator, model_summac (via summarization_metrics)
+    accelerate~=0.23.0  # For the newer versions of Transformers
```
I think `transformers` is used everywhere (for both language  and multimodal models), so it might be better to leave it here.

yifanmai replied:
`transformers` is fine but I meant `accelerate` - is that also commonly used?

teetone replied:
I have a comment but `transformers` errors without `accelerate` when running inference.

teetone replied:
I will just move it to `images`.

teetone commented:
> Partial review (sorry, didn't have time to get through all of this today).

Thanks @yifanmai! I addressed your comments. I'll follow up again once you've finished the review.

yifanmai commented on src/helm/proxy/clients/vision_language/idefics_client.py:
```diff
@@ -0,0 +1,197 @@
+        except RuntimeError as e:
+            return RequestResult(success=False, cached=False, error=str(e), completions=[], embedding=[])
+
+        completions: List[Sequence] = [Sequence(text=result["output"], logprob=0, tokens=[])]
```
Does it cause problems that `tokens` is empty?

teetone replied:
I'm not sure how to get the logprobs so I left it empty for now. I will at least tokenize and return tokens then.
